### PR TITLE
kubeadm: kube-proxy needs to know the pod subnet CIDR

### LINKED ISSUE #102: kube-proxy can't distinguish internal and external traffic
As [reported by @damaspi](https://github.com/kubernetes/kubeadm/issues/74#issuecomment-269870039):
> When I deploy some demo application, I have the same message as above. (Error syncing pod, skipping: failed to "SetupNetwork" ).
>
>When I check the logs of the proxy pod, kubectl logs kube-proxy-g7qh1 --namespace=kube-system I get the following info: proxier.go:254] clusterCIDR not specified, unable to distinguish between internal and external traffic

### PR DESCRIPTION
**Pull request by pires**
**What this PR does / why we need it**: `kube-proxy` 1.5 has a new flag `cluster-cidr` that isn't specified by `kubeadm`, thus resulting in bug https://github.com/kubernetes/kubeadm/issues/102.

**Which issue this PR fixes**: fixes https://github.com/kubernetes/kubeadm/issues/102

**Special notes for your reviewer**:
/cc @luxas @dmmcquay

### CODE REVIEW

**Comment by amacneil** [comment]
This might be a noob question, but should the kube-proxy `--cluster-cidr` be set to match the kubeadm `--pod-network-cidr` or `--service-cidr`?

**Comment by pires** [comment]
@amacneil the pod network.

**Comment by pires** [comment]
@luxas ping

**Comment by pires** [comment]
ping @luxas

**Comment by luxas** [comment]
As I mentioned on Slack, this _doesn't fix the issue in most cases._

kubeadm has _no idea_ about the pod subnet **except** when using flannel in k8s api backed mode.
In all other cases, kubeadm don't know what the CIDR is gonna be, e.g. Weave has 10.32.0.0/16 IIRC, Calico and Canal has something else, flannel defaults to 10.244.0.0/16 etc.

We just don't know this and never will. I'm curious exactly what the issue with kube-proxy is when it does not know this, I've never encountered it.

@thockin (feel free to ping somebody else that knows) why does kube-proxy have this flag? At a first glance it seems like bad design for this particular flag, but on the other hand I don't have the full context on this.

TL;DR; This PR does not solve the issue, but we should discuss how we can solve the issue anyway.

**Comment by thockin** [comment]
cmd/kube-proxy/app/options/options.go

84:	fs.StringVar(&s.ClusterCIDR, "cluster-cidr", s.ClusterCIDR, "The CIDR range of pods in the cluster. It is used to bridge traffic coming from outside of the cluster. If not provided, no off-cluster bridging will be performed.")

**Comment by pires** [comment]
AFAIK, for `kube-proxy` to properly identify internal vs external traffic, the pod subnet must be passed at all times, regardless of what the networking solution one uses, e.g. flannel or calico or manual.

**Comment by pires** [comment]
@luxas all checks are green, ping.

**Comment by luxas** [comment]
We discussed this during the sig meeting, and will do an other approach than we initially thought.

We'll encourage (and maybe even require) the user to setting this Pod CIDR value on `kubeadm init` time, and the cni daemonset yaml manifest as well. Even though it's suboptimal to start the subnet allocation feature of the controller-manager for network vendors like weave that uses something else, we chose to not expose that boolean because it would be a too detailed thing to expose to the normal user.

So we will encourage setting both the CNI network manifest and this Pod CIDR, which also allows us to do what this PR implements, namely pass that Pod CIDR to kube-proxy.

/lgtm
/approve
